BUILD.bazel

```diff
@@ -0,0 +1,268 @@
+        "src/tools/fuzzing/random.h",
+        "src/tools/js-wrapper.h",
+        "src/tools/optimization-options.h",
+        "src/tools/spec-wrapper.h",
+        "src/tools/tool-options.h",
+        "src/tools/wasm-opt.cpp",
+        "src/tools/wasm2c-wrapper.h",
+    ],
+    copts = BINARYEN_COPTS,
+    includes = ["src/tools"],
+    # copybara:strip_begin
+    visibility = ["//visibility:private"],
+    # copybara:strip_end_and_replace_begin
+    # visibility = ["//third_party/emscripten:__subpackages__"],
+    # copybara:replace_end
+    deps = [":binaryen_lib"],
+)
+
+cc_binary(
+    name = "wasm-reduce",
+    srcs = [
+        "src/tools/tool-options.h",
+        "src/tools/wasm-reduce.cpp",
+    ],
+    copts = BINARYEN_COPTS + [
+        "-Wno-unused-but-set-parameter",  # opt compiles
+    ],
+    includes = ["src/tools"],
+    deps = [":binaryen_lib"],
+)
+
+cc_binary(
+    name = "wasm-shell",
+    srcs = [
+        "src/tools/execution-results.h",
+        "src/tools/wasm-shell.cpp",
+    ],
+    copts = BINARYEN_COPTS,
+    includes = ["src/tools"],
+    deps = [":binaryen_lib"],
+)
+
+cc_binary(
+    name = "wasm-split",
+    srcs = [
+        "src/tools/tool-options.h",
+        "src/tools/wasm-split/instrumenter.cpp",
+        "src/tools/wasm-split/instrumenter.h",
+        "src/tools/wasm-split/split-options.cpp",
+        "src/tools/wasm-split/split-options.h",
+        "src/tools/wasm-split/wasm-split.cpp",
+    ],
+    copts = BINARYEN_COPTS,
+    includes = ["src/tools"],
+    deps = [":binaryen_lib"],
+)
+
+py_binary(
+    name = "generate_intrinsics",
+    srcs = ["generate_intrinsics.py"],
+    python_version = "PY3",
+    deps = ["@com_google_absl_py//absl:app"],
+)
+
+genrule(
+    name = "wasm-intrinsics",
+    srcs = [
+        "src/passes/wasm-intrinsics.wat",
+        "src/passes/WasmIntrinsics.cpp.in",
+    ],
+    outs = ["WasmIntrinsics.cpp"],
+    # copybara:strip_begin
+    cmd = "$(location :generate_intrinsics) $(SRCS) $(location :WasmIntrinsics.cpp)",
+    tools = [":generate_intrinsics"],
+    # copybara:strip_end_and_replace_begin
+    # cmd = "$(location :generate_intrinsics.par) $(SRCS) $(location :WasmIntrinsics.cpp)",
+    # tools = [":generate_intrinsics.par"],
+    # copybara:replace_end
+)
+
+# Constructs the version string in the form of:
+# version <version number> (version_<version number>-<latest commit hash>)
+# Ideally it would also include number of commits since the last version
+# was cut, but we don't easily have access to that number.
+# <version number> must contain only digits because emscripten will throw an
+# error otherwise.
+# copybara:strip_begin
+py_binary(
+    name = "generate_config",
+    srcs = ["generate_config.py"],
+    python_version = "PY3",
+    deps = ["@com_google_absl_py//absl:app"],
+)
+
+genrule(
+    name = "config",
+    outs = ["src/config.h"],
+    cmd = "$(location :generate_config) $(location :src/config.h)",
+    tools = [":generate_config"]
+)
```

Comment:
Might want to use stamping for this. See [Stamping Bazel builds with selective delivery](https://www.aspect.dev/blog/stamping-bazel-builds-with-selective-delivery)